### PR TITLE
Interceptor improvements

### LINKED ISSUE #541: [Phobos Bug] Interceptor=yes shooting down friendly projectiles
### Description:
Despite it being changed some time ago, there is still a slim chance that an Interceptor=yes object will shoot at friendly projectiles. This appears to be caused by projectile losing its previous ownership (and becoming neutral?), which makes it an eligible target to intercept.

### Steps To Reproduce:
While i suspect this will also happen when a unit fires and then dies while the projectile is in air, that can be very hard to replicate.

Instead, the easiest way to reproduce it is on a Boris-like infantry, by giving the airstrike plane a ridiculously high range (like 255, but something like 64 should also do), and then attacking your own buildings with the flare weapon. The faster the aircraft leaves the map after firing (either due to high range or just lucky positioning), the higher the chance that nearby interceptors will try to intercept the projectile.

### PR DESCRIPTION
**Pull request by Starkku**
Introduces following improvements to the projectile interceptor logic:
- Explicitly defined interceptor weapon slot via `Interceptor.Weapon`.
- Ability to customize houses projectiles from which can be intercepted via `Interceptor.CanTargetHouses`.
- Ability to use other types of projectiles than `Inviso=true` (still needs AA=yes and/or AG=yes to work depending on target projectile elevation) for interceptor weapons.
- Interceptor weapons will intercept any interceptable projectiles caught within their `CellSpread`.
- Introdued `Strength` and `Armor` for projectiles used for interceptor logic. If projectile has `Armor` set, it will only be intercepted by weapons that can target said armor, and amount equaling the weapon's `Damage` multiplied by Warhead `Verses` and the firer's firepower multipliers is deducted from the projectile's current strength which starts from `Strength`. If projectile was intercepted and its current strength is at 0 or below, regardless of if it was damaged, it will detonate.
- Introduced a toggle for critical hits to be suppressed upon interception.
- Introduced an option (both interceptable and interceptor) to not detonate intercepted projectiles and instead just delete them.
- Introdudced an option (both interceptable and interceptor) to override the intercepted projectile's weapon settings like damage and warhead.
  - Additionally on interceptors can also override projectile settings and apply damage cumulatively instead of replacing it.
- Introduced an option to keep intercepted projectiles alive instead of being destroyed.
- Fixed an issue where projectiles fired by friendly objects would be intercepted if their firer died after the projectile was launched.
- Fixed an issue with `Inviso=true` interceptor weapons detonating itself at wrong coordinates.

In addition I did some rearranging and cleaning up for hooks (most if not all weapon selection, firing etc. hooks are now in a file called `Hooks.Firing.cpp` under `TechnoClass` for an example).

Following is the documentation for the features affected:

--------------------------

### Projectile interception logic

- Projectiles can now be made interceptable by certain TechnoTypes by setting `Interceptable=true` on them. The TechnoType scans for interceptable projectiles within a range if it has no other target and will use one of its weapons to shoot at them. Note that the interceptor weapon may need `AG` and/or `AA` set to true on its projectile to be able to target projectiles depending on their elevation from ground. Projectiles can define `Armor` and `Strength`. Weapons that cannot target the projectile's armor type will not attempt to intercept it. On interception, if the projectile has `Armor` set, an amount equaling to the intercepting weapon's `Damage` adjusted by Warhead `Verses` and the TechnoType's firepower multipliers is deducted from the projectile's current strength. Regardless of if the current projectile strength was reduced or not, if it sits at 0 or below after interception, the projectile is detonated.
  - `Interceptor.Weapon` determines the weapon (0 = `Primary`, 1 = `Secondary`) to be used for intercepting projectiles.
  - `Interceptor.CanTargetHouses` controls which houses the projectiles (or rather their firers) can belong to be eligible for interception. 
  - `Interceptor.GuardRange` (and `Interceptor.(Rookie|Veteran|EliteGuardRange`) is maximum range of the unit to intercept projectile. The unit weapon range will limit the unit interception range though.
  - `Interceptor.MinimumGuardRange` (and `Interceptor.(Rookie|Veteran|EliteMinimumGuardRange`) is the minimum range of the unit to intercept projectile. Any projectile under this range will not be intercepted.
  - `Interceptable.DeleteOnIntercept` determines whether or not the projectile will simply be deleted on detonation upon interception, or if it will properly detonate. Will be overridden by `Interceptor.DeleteOnIntercept` setting on the interceptor.
  - `Interceptable.WeaponOverride` can be set to a WeaponType that will be used to override characteristics such as `Damage` and `Warhead` of the current projectile for detonation after interception. Will be overridden by `Interceptor.WeaponOverride` setting on the interceptor.
    - On interceptors, `Interceptor.WeaponReplaceProjectile` can be set to true to make `Interceptor.WeaponOverride` also replace the intercepted projectile's type (including `Image` and other projectile characteristics) and `Speed` with its own. Does not replace particle systems (`AttachedSystem`, *Ares feature*).
    - On interceptors, `Interceptor.WeaponCumulativeDamage` can be set to true to make `Damage` from `Interceptor.WeaponOverride` weapon be added on the projectile's damage rather than override it.
  - `Interceptor.KeepIntact` can be set to true to allow intercepted projectiles to continue traveling as if they were not intercepted, but effects such as `Interceptor.WeaponOverride` will still be applied.

In `rulesmd.ini`:
```ini
[SOMETECHNO]                               ; TechnoType
Interceptor=no                             ; boolean
Interceptor.Weapon=0                       ; integer, weapon slot index (0 or 1)
Interceptor.CanTargetHouses=enemies        ; Affected House Enumeration (none|owner/self|allies/ally|team|enemies/enemy|all)
Interceptor.GuardRange=0.0                 ; double
Interceptor.VeteranGuardRange=             ; double
Interceptor.EliteGuardRange=               ; double
Interceptor.MinimumGuardRange=0.0          ; double
Interceptor.VeteranMinimumGuardRange=      ; double
Interceptor.EliteMinimumGuardRange=        ; double
Interceptor.DeleteOnIntercept=false        ; boolean
Interceptor.WeaponOverride=                ; WeaponType
Interceptor.WeaponReplaceProjectile=false  ; boolean
Interceptor.WeaponCumulativeDamage=false   ; boolean
Interceptor.KeepIntact=false               ; boolean

[SOMEPROJECTILE] ; Projectile
Interceptable=false                    ; boolean
Interceptable.DeleteOnIntercept=false  ; boolean
Interceptable.WeaponOverride=          ; WeaponType
Strength=0                             ; integer
Armor=                                 ; ArmorType
```

In `rulesmd.ini`:
```ini
[SOMETECHNO]                               ; TechnoType
Interceptor=no                             ; boolean
Interceptor.Weapon=0                       ; integer, weapon slot index (0 or 1)
Interceptor.CanTargetHouses=enemies        ; Affected House Enumeration (none|owner/self|allies/ally|team|enemies/enemy|all)
Interceptor.GuardRange=0.0                 ; double
Interceptor.VeteranGuardRange=             ; double
Interceptor.EliteGuardRange=               ; double
Interceptor.MinimumGuardRange=0.0          ; double
Interceptor.VeteranMinimumGuardRange=      ; double
Interceptor.EliteMinimumGuardRange=        ; double
Interceptor.DeleteOnIntercept=false        ; boolean
Interceptor.WeaponOverride=                ; WeaponType
Interceptor.WeaponReplaceProjectile=false  ; boolean
Interceptor.WeaponCumulativeDamage=false   ; boolean
Interceptor.KeepIntact=false               ; boolean

[SOMEPROJECTILE] ; Projectile
Interceptable=false                    ; boolean
Interceptable.DeleteOnIntercept=false  ; boolean
Interceptable.WeaponOverride=          ; WeaponType
Strength=0                             ; integer
Armor=                                 ; ArmorType
```
------

### Chance-based extra damage or Warhead detonation / 'critical hits'

 - `Crit.SuppressWhenIntercepted`, if set, prevents critical hits from occuring at all if the warhead was detonated from a projectile that was intercepted.
 
  
In `rulesmd.ini`:
```ini
[SOMEWARHEAD]                       ; Warhead
Crit.SuppressWhenIntercepted=false  ; boolean
```

-------------------

Closes #541

### CODE REVIEW

**Comment by Starkku** [comment]
Further additions / changes:

- Changed GuardRange flags to use Promotable template (which means veteran-level specific guard range is now also supported). Existing flag names will still work.
- Added `Interceptor.CanTargetHouses` filter which is used instead of being hardcoded to only target enemy-fired projectiles.
- `Interceptable.DeleteOnIntercept` and `Interceptor.DeleteOnIntercept` can be used to make intercepted projectiles disappear instead of detonate. Interceptor setting takes precedence over interceptable.
- `Interceptable.WeaponOverride` and `Interceptor.WeaponOverride` can be used to override the projectile's existing weapon settings like `Damage` and `Warhead`. Interceptor setting takes precedence over interceptable.
  - `Interceptor.WeaponReplaceProjectile` can be set to also override the existing projectile's type settings such as `Image` as well as `Speed` from `Interceptor.WeaponOverride`.
  - `Interceptor.WeaponCumulativeDamage`, if set, makes damage from `Interceptor.WeaponOverride` add upon existing damage instead of overriding it.
- `Interceptor.KeepIntact` if set, keeps intercepted projectiles intact and does not detonate or delete them.

File: docs/New-or-Enhanced-Logics.md [hunk 1]
```diff
@@ -644,6 +662,7 @@ RemoveMindControl=no                 ; boolean
   - `Crit.AffectsBelowPercent` can be used to set minimum percentage of their maximum `Strength` that targets must have left to be affected by a critical hit.
   - `Crit.AnimList` can be used to set a list of animations used instead of Warhead's `AnimList` if Warhead deals a critical hit to even one target. If `Crit.AnimList.PickRandom` is set (defaults to `AnimList.PickRandom`) then the animation is chosen randomly from the list.
     - `Crit.AnimOnAffectedTargets`, if set, makes the animation(s) from `Crit.AnimList` play on each affected target *in addition* to animation from Warhead's `AnimList` playing as normal instead of replacing `AnimList` animation.
+  - `Crit.SuppressOnIntercept`, if set, prevents critical hits from occuring at all if the projectile the warhead was detonated from a [projectile that was intercepted](#projectile-interception-logic).
```
Comment (Metadorius) [inline docs/New-or-Enhanced-Logics.md:665]:
```suggestion
  - `Crit.SuppressOnIntercept`, if set, prevents critical hits from occuring at all if the warhead was detonated from a [projectile that was intercepted](#projectile-interception-logic).
```
Also I suggest to name it `Crit.SuppressWhenIntercepted` to denote clearly that crits won't work when *intercepted*, not *intercepting*.

File: src/Misc/Hooks.BugFixes.cpp [hunk 2]
```diff
@@ -370,7 +370,7 @@ DEFINE_HOOK(0x480552, CellClass_AttachesToNeighbourOverlay_Gate, 0x7)
 	return 0;
 }
 
-DEFINE_HOOK(0x415F5C, AircraftClass_FireAt_SpeedModifiers, 0xA)
+DEFINE_HOOK(0x415F5C, AircraftClass_Fire_At_SpeedModifiers, 0xA)
```
[the remark this answers is not in the document]
Comment (Metadorius) [reply src/Misc/Hooks.BugFixes.cpp:373]: Yeah I think we're keeping to no underscore style now, would be good to use it IMO.

**Comment by ayylmaoRotE** [comment]
Tested all tags and can confirm they work, even LaserLineTrails get applied to the projectile if replaced when intercepted. It's worth noting only `ROT=x` will continue its path to the target, all other types will explode on the spot (similar to airburst)

There's 1 big issue though, only `ROT=x` and `Arcing=yes` projectiles can be intercepted. 
Projectiles using `Trajectory=Straight` or `Gravity=0/Gravity.HeightFix=true` do not get targeted at all.
image


The tests were done with an interceptable projectile using custom armor: `missile=0%` and interceptor using 0% vs everything and `Versus.missile=100%`

The interceptor targeting could use some work though.. several missiles fired at my Paladins and they all decide to intercept a single one.
image

Also, having multiple WeaponOverrides doesn't appear to do anything crazy and only one weapon is selected 👍 I only mention this since they all target the same projectile simultaneously

**Comment by Starkku** [comment]
Cannot reproduce the issue where `Trajectory` projectiles are uninterceptable. They are intercepted just fine on my end, only that they tend to be so fast that interceptors cannot often react timely.

I changed the interceptor target scan to prioritize untargeted projectiles. If there are none, it will pick whatever it found first just like before though.

**Comment by Starkku** [comment]
Clarified interaction of `AA`/`AG` on interceptor weapon projectiles in documentation.

File: docs/New-or-Enhanced-Logics.md [hunk 3]
```diff
@@ -323,23 +323,41 @@ Ammo.Shared.Group=-1                    ; integer
...
-Interceptor.EliteGuardRange=0.0         ; double
-Interceptor.MinimumGuardRange=0.0       ; double
-Interceptor.EliteMinimumGuardRange=0.0  ; double
+- Projectiles can now be made interceptable by certain TechnoTypes by setting `Interceptable=true` on them. The TechnoType scans for interceptable projectiles within a range if it has no other target and will use one of its weapons to shoot at them. Note that the interceptor weapon may need `AG` and/or `AA` set to true on its projectile to be able to target projectiles depending on their elevation from ground. Projectiles can define `Armor` and `Strength`. Weapons that cannot target the projectile's armor type will not attempt to intercept it. On interception, if the projectile has `Armor` set, an amount equaling to the intercepting weapon's `Damage` adjusted by Warhead `Verses` and the TechnoType's firepower multipliers is deducted from the projectile's current strength. Regardless of if the current projectile strength was reduced or not, if it sits at 0 or below after interception, the projectile is detonated.
+  - `Interceptor.Weapon` determines the weapon (0 = `Primary`, 1 = `Secondary`) to be used for intercepting projectiles.
```
Comment (Metadorius) [inline docs/New-or-Enhanced-Logics.md:327]:
```suggestion
- Projectiles can now be made interceptable by certain TechnoTypes by setting `Interceptable=true` on them. The TechnoType scans for interceptable projectiles within a range if it has no other target and will use one of its weapons to shoot at them. Projectiles can define `Armor` and `Strength`. Weapons that cannot target the projectile's armor type will not attempt to intercept it. On interception, if the projectile has `Armor` set, an amount equaling to the intercepting weapon's `Damage` adjusted by Warhead `Verses` and the TechnoType's firepower multipliers is deducted from the projectile's current strength. Regardless of if the current projectile strength was reduced or not, if it sits at 0 or below after interception, the projectile is detonated.
  - `Interceptor.Weapon` determines the weapon (0 = `Primary`, 1 = `Secondary`) to be used for intercepting projectiles.
    - The interceptor weapon may need `AG` and/or `AA` set to true on its projectile to be able to target projectiles depending on their elevation from ground. If you don't set those then the weapon won't be able to target low-flying or high-flying projectiles respectively.
```

**Comment by mevitar** [comment]
Didn't notice any issues, but as with other weapon calls, Interceptable.WeaponOverride and Interceptor.WeaponOverride might require the weapon to be specified under [WeaponTypes]. I've had random weapons not work when used as an override and tried to figure out why, until i remembered that the list exists, and putting the weapons there obviously fixed it.

**Review by Metadorius — APPROVED** [review]

**Pull request merged**